add context for unit and e2e testing (#245)
pxblue/doc-it: src/docs/development/testing.mdx

-   **Unit Tests**: verify individual methods, functions, or components.
-   **Integration Tests**: verify that different modules or services in your application work together.
-   **Functional Tests**: verify the output of certain actions based on business requirements.
-   **End-to-end (E2E) Tests**: verify that user behaviors / workflows function as expected.
-   **Acceptance Tests**: verify that the entire system meets the business requirements.
-   **Performance Tests**: verify that the system behaves correctly under heavy load.
-   **Smoke Tests: quick**: basic checks of the major features of the application.

This document focuses mainly on Unit Tests and E2E tests, as they are the easiest to automate.

<Divider />

# Unit Tests

Unit tests target the smallest pieces of your application (such as verifying inputs and outputs of a class method or function). These will likely make up the bulk of your tests because they are quick to write, easy to automate, and can help you achieve significant code coverage.

> Code Coverage measures the percentage of your code that is verified by your tests.

> A unit test shouldn't have any external dependencies, such as other methods or APIs. By keeping unit tests isolated in this manner, it's easy to identify the cause of a failing test an implement a fix quickly. When you start combining features and methods together, you move into the realm of _integration testing_.

Reviewer: ```suggestion
A unit test shouldn't have any external dependencies, such as other methods or APIs. By keeping unit tests isolated in this manner, it's easy to identify the cause of a failing test and implement a fix quickly. When you start combining features and methods together, you move into the realm of _integration testing_.
```

The most important thing to keep in mind about tests is that they should provide confidence that the code and design are working as intended without issues in any scenario. For most teams, it's fine to write tests after features and functions are developed, so long as you write quality tests and achieve good coverage.

## How to Write

When unit testing components, the most effective tests are based on what is actually rendered on the screen rather than the internal implementation logic. The test should consider the application from the user's perspective and test things the user views and interacts with.

For example, consider testing an Open Button that should open a modal dialog when clicked. Your test should check that a button with the _Open_ label is rendered on the screen, and when it is clicked, the dialog is rendered on the screen with the correct content. You may be tempted to identify the button by a class or id, or test that a particular function is called when the button is pressed - these tests are not as effective because they are more likely to break if the implementation of the component changes, even though the end behavior is the same.

In general:

> -   Test what the user sees (i.e., rendered output) and what the user can do (i.e., interactions).

Reviewer: ```suggestion
-   Test what the user sees (i.e., rendered output) and what the user can do  (i.e., interactions).
```

Behavior Driven Development “BDD” methodology unit test are written before any implementation of code for a feature and after the feature story has been written.

The methodologies are not always followed due to type of feature or application being developed but instead the unit test outcome should provide confidence that the code and design are working as intended without issues in any scenario. 

Write unit tests in a way that they test the render logic and not the internal implementation. The test should be testing the things that the user will be viewing and interacting with.

Let’s say you are testing a component that displays a button on the screen and when the user clicks the button, your component renders a message in a div element.

Your test should check practically for the same,  check if the button is clickable and if the div exists after clicking the button.

Example component with a button.

- We should test what the user views i.e. rendering and what the user can do i.e. interactions.

- If there is visual change after clicking , test for it.

- If there is a value returned after clicking, test for it.

> ```sh

Reviewer 2: ```suggestion
```tsx
```

For example, consider testing an Open Button that should open a modal dialog when clicked. Your test should check that a button with the _Open_ label is rendered on the screen, and when it is clicked, the dialog is rendered on the screen with the correct content. You may be tempted to identify the button by a class or id, or test that a particular function is called when the button is pressed - these tests are not as effective because they are more likely to break if the implementation of the component changes, even though the end behavior is the same.

> -   Test what the user sees (i.e., rendered output) and what the user can do (i.e., interactions).

Reviewer: missing a space here
```suggestion
-   Test what the user sees (i.e., rendered output) and what the user can do  (i.e., interactions).
```

Functional tests alone will not give you enough test coverage to avoid regressions. Unit tests aim to provide code coverage depth, while functional tests provide coverage over the breadth of user test scenarios.

## How to Write

End-to-end tests rely on the ability to find elements on the screen and interact with them in an automated way. This means that you need a reliable mechanism for selecting elements. Depending on your testing framework, there are different ways to select elements, including:

-   **CSS Selectors**: find elements by using CSS classes or rules (e.g., `.your-class-name`)
-   **Element Selectors**: find elements inside the elements (e.g., `#\@\@pxb-drawerlayout-content > div > header > div > button`)
-   **xPath Selectors**: find selector within xPath (e.g., `//*[@id="@@pxb-drawerlayout-content"]/div/header/div/button`)
-   **ID Selectors**: find attribute id within elements (e.g., `#login-button`)

> See selector strategies on [browserstack](https://www.browserstack.com/guide/css-selectors-in-selenium) and [software testing help](https://www.softwaretestinghelp.com/css-selector-selenium-locator-selenium-tutorial-6/)

Reviewer: ```suggestion
> You can read more about different selector strategies on [browserstack](https://www.browserstack.com/guide/css-selectors-in-selenium) and [software testing help](https://www.softwaretestinghelp.com/css-selector-selenium-locator-selenium-tutorial-6/).
```

One way to combat this in your own applications is to use dedicated attributes for testing (e.g., `test-id` or `dev-id`). By adding this test-specific attribute to the element, identifying the correct underlying component should be safer. Because this attribute is test-specific and the application logic does not use it, it will be less likely to change throughout the product lifecycle. Combining this approach with other selectors is a common and reliable approach for element identification.

```html
<button test-id="login-button" class="default-button">Login</button>
```

Selector to use in a test

```
WebElement click = driver.findElement(By.id("login-button"));

> ```

Reviewer: Does by.id work here if the attribute is test-id?

Reviewer: ```suggestion
```html
// Component code
<button test-id="login-button" class="default-button">Login</button>

// Selector to use in a test
WebElement click = driver.findElement(By.id("login-button"));
```
```

## Testing Frameworks

There are many tools available for configuring automated E2E tests. Some of the more popular tools include:

-   Selenium
-   Cypress

> -   WebDriverIO

Reviewer: Do these have a main documentation page that we could link to here?

### Selenium

> Selenium is a popular automation testing suite which can be used to automate the desktop and mobile web browser interactions. You can perform automation testing by writing code in any of your preferred language supported by Selenium.

Reviewer: ```suggestion
Selenium is a popular automation testing suite which can be used to automate the desktop and mobile web browser interactions. You can write test code in any of the languages supported by Selenium.
```

> Selenium test suite has several test frameworks available and each one can be custmized to your project needs.

Reviewer: ```suggestion
The Selenium test suite has several test frameworks available and each one can be customized to your project needs.
```

> For more information see [browserstack](https://www.browserstack.com/guide/selenium-framework) selenium guide or [selenium on github](https://github.com/SeleniumHQ/selenium) documentation.

Reviewer: ```suggestion
For more information, check out the browserstack [selenium guide](https://www.browserstack.com/guide/selenium-framework) or the official [selenium documentation](https://github.com/SeleniumHQ/selenium.
```

> For selenium tutorial series see [software testing help](https://www.softwaretestinghelp.com/selenium-tutorial-1/)

Reviewer: ```suggestion
There are also a number of [Selenium tutorials](https://www.softwaretestinghelp.com/selenium-tutorial-1/) available online to help you get started.
```

### Cypress

Cypress is a popular open source testing framework that boasts support for any modern JavaScript framework. It works well for E2E tests in both Angular and React applications and is easier to use than some of the more traditional tools, such as Selenium.

For a more detailed write-up on Cypress, check out the official [Cypress Documentation](https://docs.cypress.io/guides/getting-started/installing-cypress.html#System-requirements).

### WebDriverIO

> WebdriverIO is a popular Javascript based test automation framework built over nodeJs. It is an open-source project developed for the automation testing community. WebdriverIO is extendible, compatible and feature-rich.

Reviewer: ```suggestion
WebdriverIO is a popular Javascript based test automation framework built on top of node.js. It is an open-source project developed for the automation testing community. WebdriverIO is extendible, compatible and feature-rich.
```

> For more information see [webdriver.io](https://webdriver.io/) and [webdriver.io](https://webdriver.io/docs/gettingstarted.html) getting started.

Reviewer: ```suggestion
For more information, check out the official [Webdriver.io documentation](https://webdriver.io/) and their [getting started guide](https://webdriver.io/docs/gettingstarted.html).
```

<Divider />

# Mobile Device Testing

> Coming soon...

Reviewer: Should we add something here to talk about the different options that are available (at a high level)? Or just leave this our for now until later?

One way to combat this in your own applications is to use dedicated attributes for testing (e.g., `test-id` or `dev-id`). By adding this test-specific attribute to the element, identifying the correct underlying component should be safer. Because this attribute is test-specific and the application logic does not use it, it will be less likely to change throughout the product lifecycle. Combining this approach with other selectors is a common and reliable approach for element identification.

```html
// Component code

<button id="login-button" class="default-button">Login</button>

> // Selector to use in a test WebElement click = driver.findElement(By.id("login-button"));

Reviewer: looks like prettier may have messed this code block up

Author: think its better now
